### PR TITLE
CoreML export and import for models with one-hot encoding features

### LINKED ISSUE #694: Support one hot categorical features in CoreML model export
CatBoost allows to save models for [Apple's CoreML format](https://developer.apple.com/documentation/coreml) but only for models without categorical features.

The goal is to support saving models that use categorical features of one hot variety (but not [counter variety](https://tech.yandex.com/catboost/doc/dg/concepts/algorithm-main-stages_cat-to-numberic-docpage/), CoreML format is not currently designed to be able to accomodate them).

CoreML models are stored as files with serialized protobuf messages. See format description [here](https://apple.github.io/coremltools/coremlspecification/)

Code that exports models to CoreML format is called [here](https://github.com/catboost/catboost/blob/496c3fecb66350badca089ac3709b1a604db5c65/catboost/libs/model/model.cpp#L182-L190).

Code that loads CoreML models is called [here](https://github.com/catboost/catboost/blob/496c3fecb66350badca089ac3709b1a604db5c65/catboost/libs/model/model.cpp#L85-L88).

This functionality is tested by python package's tests: [1](https://github.com/catboost/catboost/blob/871d215df4376c6a25d5c83ab5797088d09dd66b/catboost/python-package/ut/medium/test.py#L703-L714), [2](https://github.com/catboost/catboost/blob/871d215df4376c6a25d5c83ab5797088d09dd66b/catboost/python-package/ut/medium/test.py#L738-L755).

### PR DESCRIPTION
**Pull request by infected-mushroom**
Fixes #694
Added saving models with one-hot encoding features to CoreML model format and import of these models to CBM format.

### CODE REVIEW

File: catboost/libs/model/coreml_helpers.cpp [hunk 1]
```diff
@@ -2,20 +2,44 @@
 #include "model_build_helper.h"
```
Comment (annaveronika) [inline catboost/libs/model/coreml_helpers.cpp]: Please add categorical features support to CoreML tutorial. Use titanic dataset for it (catboost.datasets.titanic())

Comment (annaveronika) [reply catboost/libs/model/coreml_helpers.cpp]: it is in catboost/tutorials/apply_model/core_ml

File: catboost/libs/model/coreml_helpers.cpp [hunk 2]
```diff
@@ -168,15 +321,31 @@ void ProcessOneTree(const TVector<const TreeEnsembleParameters::TreeNode*>& tree
...
+                }
             } else {
-                auto& floatSplit = splits->at(nodeLayerIds[nodeId]);
-                CB_ENSURE(floatSplit.FloatFeature == (int)node->branchfeatureindex());
-                CB_ENSURE(floatSplit.Split == node->branchfeaturevalue());
+                if (splits->ysize() <= nodeLayerIds[nodeId]) {
+                    splits->resize(nodeLayerIds[nodeId] + 1);
+                    auto& split = splits->at(nodeLayerIds[nodeId]);
+                    split.OneHotFeature.CatFeatureIdx = node->branchfeatureindex();
+                    split.OneHotFeature.Value = (int)node->branchfeaturevalue();
```
Comment (andrey-khropov) [inline catboost/libs/model/coreml_helpers.cpp]: Please, add a comment that Value is not hashed value as it should be and the field is used temporarily as an index in the OneHotFeature Values array with hashed values.

File: catboost/libs/model/coreml_helpers.cpp [hunk 3]
```diff
@@ -188,8 +357,34 @@ void ProcessOneTree(const TVector<const TreeEnsembleParameters::TreeNode*>& tree
 
 void NCatboost::NCoreML::ConvertCoreMLToCatboostModel(const Model& coreMLModel, TFullModel* fullModel) {
     CB_ENSURE(coreMLModel.specificationversion() == 1, "expected specificationVersion == 1");
-    CB_ENSURE(coreMLModel.has_treeensembleregressor(), "expected treeensembleregressor model");
-    auto& regressor = coreMLModel.treeensembleregressor();
+    TreeEnsembleRegressor regressor;
+    THashMap<int, THashMap<int, TString>> categoricalMappings;
```
Comment (andrey-khropov) [inline catboost/libs/model/coreml_helpers.cpp]: Please, add a comment about the meaning of keys and values in this mapping.

File: catboost/libs/model/coreml_helpers.cpp [hunk 4]
```diff
@@ -188,8 +357,34 @@ void ProcessOneTree(const TVector<const TreeEnsembleParameters::TreeNode*>& tree
...
+        CB_ENSURE(modelsCount > 0, "expected nonempty pipeline");
+        CB_ENSURE(models.Get(modelsCount - 1).has_treeensembleregressor(), "expected treeensembleregressor model");
+        regressor = models.Get(modelsCount - 1).treeensembleregressor();
+
+        for (size_t i = 0; i < modelsCount - 1; i++) {
+            auto& model = models.Get(i);
+            CB_ENSURE(model.has_categoricalmapping(), "expected categorical mappings in pipeline");
+
+            auto& mapping = model.categoricalmapping().stringtoint64map().map();
+            int featureName = std::stoi(model.description().input(0).name().substr(8));
```
Comment (andrey-khropov) [inline catboost/libs/model/coreml_helpers.cpp]: Comment about why 'substr(8)' is needed.

File: catboost/libs/model/coreml_helpers.cpp [hunk 5]
```diff
@@ -188,8 +357,34 @@ void ProcessOneTree(const TVector<const TreeEnsembleParameters::TreeNode*>& tree
...
+            THashMap<int, TString> invertMapping;
```
Comment (andrey-khropov) [inline catboost/libs/model/coreml_helpers.cpp]: invertMapping -> invertedMapping. It is a variable, not a function.

File: catboost/libs/model/coreml_helpers.cpp [hunk 6]
```diff
@@ -198,45 +393,87 @@ void NCatboost::NCoreML::ConvertCoreMLToCatboostModel(const Model& coreMLModel,
     int approxDimension = ensemble.numpredictiondimensions();
 
     TVector<TVector<const TreeEnsembleParameters::TreeNode*>> treeNodes;
-    TVector<TVector<TFloatSplit>> trees;
+    TVector<TVector<TModelSplit>> treesSplits;
```
Comment (andrey-khropov) [inline catboost/libs/model/coreml_helpers.cpp]: Comment about different fields meaning (FloatFeature.FloatFeature, OneHotFeature.CatFeatureIdx, OneHotFeature.Value) when filling this structure in ProcessOneTree is necessary.

**Comment by andrey-khropov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@andrey-khropov, internal review request created: [825528](https://a.yandex-team.ru/review/825528)

**Comment by arcadia-devtools** [comment]
Check [825528](https://a.yandex-team.ru/review/825528) failed: 1 test(s) broken

**Pull request merged**